File: typescript/src/agents/citationAgent.ts
import type { ChatModel } from "../providers/types.js";
import type { Harness } from "../harness/index.js";
import type { ToolRuntime } from "../tools/types.js";
import type { Tracer } from "../tracing/tracer.js";
import { runAgent } from "./runAgent.js";
import type { AgentResult } from "./types.js";

export interface RunCitationAgentParams {
  findings: AgentResult[];
  model: ChatModel;
  harness: Harness;
  runtime: ToolRuntime;
  tracer: Tracer;
  runId: string;
  parentSpanId?: string | null;
}

export async function runCitationAgent(params: RunCitationAgentParams): Promise<AgentResult> {
  const findingsText = params.findings
    .map((f) => `[${f.taskId}] (${f.status}) ${f.text}` + (f.artifactRefs.length ? ` refs: ${f.artifactRefs.map((r) => r.artifactId).join(", ")}` : ""))
    .join("\n\n");

  return runAgent({
    roleName: "citation",
    userPrompt: findingsText,
    model: params.model,
    harness: params.harness,
    runtime: params.runtime,
    tracer: params.tracer,
    taskId: `${params.runId}-citation`,
    delegationDepth: 0,
    parentSpanId: params.parentSpanId ?? null,
  });
}

File: python/src/multiagent_boilerplate/agents/citation_agent.py
from __future__ import annotations

from ..harness import Harness
from ..providers.types import ChatModel
from ..tools.types import ToolRuntime
from ..tracing.tracer import Tracer
from .run_agent import RunAgentParams, run_agent
from .types import AgentResult


async def run_citation_agent(
    findings: list[AgentResult],
    model: ChatModel,
    harness: Harness,
    runtime: ToolRuntime,
    tracer: Tracer,
    run_id: str,
    parent_span_id: str | None = None,
) -> AgentResult:
    findings_text = "\n\n".join(
        f"[{f.task_id}] ({f.status}) {f.text}"
        + (f" refs: {', '.join(r['artifactId'] for r in f.artifact_refs)}" if f.artifact_refs else "")
        for f in findings
    )

    return await run_agent(
        RunAgentParams(
            role_name="citation",
            user_prompt=findings_text,
            model=model,
            harness=harness,
            runtime=runtime,
            tracer=tracer,
            task_id=f"{run_id}-citation",
            delegation_depth=0,
            parent_span_id=parent_span_id,
        )
    )
